source/module_base/para_json.cpp

```diff
@@ -0,0 +1,986 @@
+
+    // @param init -- Kpoints
+    rapidjson::Value kpoints(rapidjson::kObjectType);
+    rapidjson::Value nkstot;
+    rapidjson::Value nkstot_ibz;
+    rapidjson::Value coordinates(rapidjson::kArrayType);
+    rapidjson::Value weight(rapidjson::kArrayType);
+
+    // @param init -- grid
+    rapidjson::Value grid(rapidjson::kObjectType);
+    rapidjson::Value energy_cutoff_for_wavefunc;
+    rapidjson::Value fft_grid_for_wave_functions(rapidjson::kArrayType);
+    rapidjson::Value number_of_plane_waves;
+    rapidjson::Value number_of_sticks;
+
+    // @param init -- Smearing
+    // rapidjson::Value smearing_method;
+    // rapidjson::Value smearing_sigma;
+
+    // @param init -- mixing
+    rapidjson::Value mixing;
+
+
+    // @param output
+    rapidjson::Value output(rapidjson::kArrayType);
+
+
+
+    // @param final_stru
+    rapidjson::Value final_stru(rapidjson::kObjectType);
+    rapidjson::Value cell;
+    rapidjson::Value coordinate;
+
+
+
+
+    /**
+     *  The functions below initialize the json output parameter 
+     *  tree to connect the nodes of the module
+    */
+
+    /**
+     * @brief   add Top stage：parameter in Abacus:
+     */
+    void Init_json_abacus()
+    {
+
+
+        // add First stage：parameter in abcus:
+
+        abacus.AddMember("general_info", general_info, doc.GetAllocator());
+
+        abacus.AddMember("readin_info", readin_info, doc.GetAllocator());
+        
+        abacus.AddMember("init", init, doc.GetAllocator());
+
+        abacus.AddMember("output", output, doc.GetAllocator());
+
+        abacus.AddMember("final_stru", final_stru, doc.GetAllocator());
+
+        doc.SetObject();
+        // abacus.SetObject();
+        doc.AddMember("ABACUS", abacus, doc.GetAllocator());
+        /**
+         * .
+         * .
+         * .
+         * .
+         * .
+         * .
+         * .
+         * */
+    }
+    /**
+     * @brief   add Second stage：parameter in Abacus - general_info:
+     */
+    void Init_json_abacus_generalInfo(){
+        general_info.AddMember("version", version, doc.GetAllocator());
+
+        general_info.AddMember("commit", commit, doc.GetAllocator());      
+
+        general_info.AddMember("begin_time", begin_time, doc.GetAllocator());      
+
+        general_info.AddMember("begin_date", begin_date, doc.GetAllocator());     
+
+        general_info.AddMember("device_g", device_g, doc.GetAllocator());                
+
+        // add Third stage：parameter in parallel:
+        general_info.AddMember("parallel", parallel, doc.GetAllocator());
+        
+        parallel.AddMember("drank", drank, doc.GetAllocator());
+
+        parallel.AddMember("dsize", dsize, doc.GetAllocator());
+                        
+        parallel.AddMember("dcolor", dcolor, doc.GetAllocator());
+                
+    }
+
+
+
+    /**
+     * @brief   add Second stage：parameter in Abacus - readin_info:
+     */
+    void Init_json_abacus_readinInfo(){
```

Comment:
Here, do not need all input parameters, but only the parameters writen in INPUT file that is supplied by user.
And, do not add an extra layer, just AddMember for readin(?)

Reply:
The former can be achieved through a non-empty detection, the latter question I don't think is a good design, there are a lot of variables in input_file, I think the module storage structure will be much clearer, if each module needs to add parameters then it will be easy to modify the code.

Reply:
In the INPUT file, the parameters are not classified to different modules. Here we just record the parameters defined in INPUT file, but not need all parameters. The purpose is just to record what parameters defined in this calculation.